### PR TITLE
Refactor KCP watcher reconcile loop

### PR DESCRIPTION
**Pull request by khlifi411**
### Description:

- Refactor tests and implement web-hook removal.
- Remove `util` package.
- Introduce `custom` package with a custom Istio client witch manages operations on virtual service resource.
- Delegate Virtual service and Gateway resources management to the KCP-watcher deployment (i.e. SREs pipeline)
- Refactor `deploy` package and remove duplicate functions.

### Related issues:
#67

### CODE REVIEW

File: kcp/controllers/watcher_controller.go [hunk 1]
```diff
@@ -141,15 +149,21 @@ func (r *WatcherReconciler) HandleProcessingState(ctx context.Context,
...
+	if err != nil {
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to delete SKR config")
 	}
 	updated := controllerutil.RemoveFinalizer(obj, watcherFinalizer)
 	if !updated {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to remove finalizer")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to remove finalizer")
 	}
 	err = r.Update(ctx, obj)
 	if err != nil {
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:169]: please correct error handling

File: kcp/controllers/watcher_controller.go [hunk 2]
```diff
@@ -159,181 +173,58 @@ func (r *WatcherReconciler) HandleDeletingState(ctx context.Context, logger logr
...
-		return apiErr
-	}
-	return err
+func addReadyCondition(obj *watcherv1alpha1.Watcher, state watcherv1alpha1.WatcherConditionStatus, msg string) {
+	obj.Status.Conditions = append(obj.Status.Conditions, watcherv1alpha1.WatcherCondition{
+		Type:               watcherv1alpha1.ConditionTypeReady,
+		Status:             state,
+		Message:            msg,
+		LastTransitionTime: &metav1.Time{Time: time.Now()},
+	})
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:213]: this appends all the time, there should be only one Ready condition per resource

File: kcp/controllers/watcher_controller.go [hunk 3]
```diff
@@ -101,38 +106,41 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+	err = deploy.UpdateWebhookConfig(ctx, r.Config.WebhookChartPath, r.Config.WebhookChartReleaseName, obj,
+		r.RestConfig, r.Client)
 	if err != nil {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to update SKR config")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to update SKR config")
 	}
-	err = r.updateWatcherCRStatus(ctx, obj, componentv1alpha1.WatcherStateReady, "successfully reconciled watcher cr")
+	err = r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateReady, "successfully reconciled watcher cr")
 	if err != nil {
 		logger.Error(err, "failed to update watcher cr to ready status")
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:145]: another one

**Comment by adityabhatia** [comment]
Think about moving /pkg to /internal

File: kcp/config/manager/istio-resources.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  name: kcp-events
+  namespace: system
+  labels:
+    kcp-watcher-resource: "true"
+spec:
+  gateways:
+    - kcp-gateway
+  hosts:
+    - "*"
+---
```
Comment (adityabhatia) [inline kcp/config/manager/istio-resources.yaml:29]: please remove as discussed

File: kcp/controllers/assets/istio-test-resources.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        number: 80
+        protocol: HTTP
+---
+apiVersion: networking.istio.io/v1beta1
+kind: VirtualService
+metadata:
+  name: kcp-events
+  namespace: default
+  labels:
+    kcp-watcher-resource: "true"
```
Comment (adityabhatia) [inline kcp/controllers/assets/istio-test-resources.yaml:23]: does this label have a significance? If not please remove to avoid confusion

Comment (khlifi411) [reply kcp/controllers/assets/istio-test-resources.yaml]: It doesn't have any. Deleted 👍

File: kcp/config/manager/istio-resources.yaml (hunk 4 above)
Comment (khlifi411) [reply kcp/config/manager/istio-resources.yaml]: Deleted 👍

File: kcp/controllers/watcher_controller.go [hunk 6]
```diff
@@ -101,239 +105,120 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 	if !updated {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to remove finalizer")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to remove finalizer")
 	}
 	err = r.Update(ctx, obj)
 	if err != nil {
-		logger.Error(err, "failed to update watcher cr")
+		msg := "failed to update watcher cr"
+		logger.Error(err, msg)
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, msg)
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:173]: this way no error will be returned!

File: kcp/controllers/watcher_controller.go [hunk 7]
```diff
@@ -101,239 +105,120 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 func (r *WatcherReconciler) HandleProcessingState(ctx context.Context,
-	logger logr.Logger, obj *componentv1alpha1.Watcher,
+	logger logr.Logger, obj *watcherv1alpha1.Watcher,
 ) error {
-	err := r.createOrUpdateServiceMeshConfigForCR(ctx, obj)
+
+	err := r.UpdateVirtualServiceConfig(ctx, r.Config.VirtualServiceObjKey, obj)
 	if err != nil {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to create or update service mesh config")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError,
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:134]: return original error if update returns no error

File: kcp/controllers/watcher_controller.go [hunk 8]
```diff
@@ -101,239 +105,120 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+			"failed to create or update service mesh config")
 	}
-	err = r.updateSKRWatcherConfigForCR(ctx, obj)
+	err = deploy.UpdateWebhookConfig(ctx, r.Config.WebhookChartPath, r.Config.WebhookChartReleaseName, obj,
+		r.RestConfig, r.Client)
 	if err != nil {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to update SKR config")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to update SKR config")
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:140]: same as above

File: kcp/controllers/watcher_controller.go [hunk 9]
```diff
@@ -101,239 +105,120 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 	}
-	err = r.updateWatcherCRStatus(ctx, obj, componentv1alpha1.WatcherStateReady, "successfully reconciled watcher cr")
+	err = r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateReady, "successfully reconciled watcher cr")
 	if err != nil {
-		logger.Error(err, "failed to update watcher cr to ready status")
+		msg := "failed to update watcher cr to ready status"
+		logger.Error(err, msg)
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, msg)
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:146]: same

File: kcp/controllers/watcher_controller.go [hunk 10]
```diff
@@ -101,239 +105,120 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+	err := r.RemoveVirtualServiceConfigForCR(ctx, r.Config.VirtualServiceObjKey, obj)
 	if err != nil {
-		return r.updateWatcherCRErrStatus(ctx, logger, err, obj, "failed to delete service mesh config")
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError,
+			"failed to delete service mesh config")
+	}
+	err = deploy.RemoveWebhookConfig(ctx, r.Config.WebhookChartPath, r.Config.WebhookChartReleaseName, obj,
+		r.RestConfig, r.Client)
+	if err != nil {
+		return r.updateWatcherCRStatus(ctx, obj, watcherv1alpha1.WatcherStateError, "failed to delete SKR config")
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:163]: same as above

**Review by adityabhatia — CHANGES_REQUESTED** [review]
Thanks for fixing all the issues. 
We can take the discussion of the fake istio clients to another PR, as you suggested.
But the update error comments should be fixed as part of this PR.

File: kcp/controllers/watcher_controller.go [hunk 11]
```diff
@@ -101,239 +105,143 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
-	state componentv1alpha1.WatcherState, msg string,
+func (r *WatcherReconciler) updateWatcherCRStatus(ctx context.Context, obj *watcherv1alpha1.Watcher,
+	state watcherv1alpha1.WatcherState, msg string,
 ) error {
 	obj.Status.State = state
 	switch state { //nolint:exhaustive
-	case componentv1alpha1.WatcherStateReady:
-		util.AddReadyCondition(obj, componentv1alpha1.ConditionStatusTrue, msg)
+	case watcherv1alpha1.WatcherStateReady:
+		obj.AddOrUpdateReadyCondition(watcherv1alpha1.ConditionStatusTrue, msg)
```
Comment (adityabhatia) [inline kcp/controllers/watcher_controller.go:224]:
```suggestion
func (r *WatcherReconciler) updateWatcherCRStatus(ctx context.Context, obj *watcherv1alpha1.Watcher,
	state watcherv1alpha1.WatcherState, msg string, sourceErr error
) error {
	obj.Status.State = state
	switch state { //nolint:exhaustive
	case watcherv1alpha1.WatcherStateReady:
		obj.AddOrUpdateReadyCondition(watcherv1alpha1.ConditionStatusTrue, msg)
	case "":
		obj.AddOrUpdateReadyCondition(watcherv1alpha1.ConditionStatusUnknown, msg)
	default:
		obj.AddOrUpdateReadyCondition(watcherv1alpha1.ConditionStatusFalse, msg)
	}
	if updateErr := r.Status().Update(ctx, obj.SetObservedGeneration()); updateErr != nil {
	         return updateErr
	}
	return sourceErr
}
```

Comment (adityabhatia) [reply kcp/controllers/watcher_controller.go:224]: I will do this in my upcoming PR

**Review by adityabhatia — APPROVED** [review]

**Pull request merged**